src/ipmi.c

```diff
@@ -682,166 +815,410 @@ static void domain_connection_change_handler(ipmi_domain_t *domain, int err,
+  status = c_ipmi_thread_init(st);
   if (status != 0) {
-    ERROR("ipmi plugin: thread_init failed.\n");
+    ERROR("ipmi plugin: c_ipmi_thread_init failed.");
+    st->active = 0;
     return (void *)-1;
   }
 
-  while (c_ipmi_active != 0) {
+  while (st->active != 0) {
     struct timeval tv = {1, 0};
     os_handler->perform_one_op(os_handler, &tv);
   }
+  return (void *)0;
+} /* void *c_ipmi_thread_main */
 
-  ipmi_posix_thread_free_os_handler(os_handler);
+static c_ipmi_instance_t *c_ipmi_init_instance() {
+  c_ipmi_instance_t *st;
 
-  return (void *)0;
-} /* void *thread_main */
-
-static int c_ipmi_config(const char *key, const char *value) {
-  if (ignorelist == NULL)
-    ignorelist = ignorelist_create(/* invert = */ 1);
-  if (ignorelist == NULL)
-    return 1;
-
-  if (strcasecmp("Sensor", key) == 0) {
-    ignorelist_add(ignorelist, value);
-  } else if (strcasecmp("IgnoreSelected", key) == 0) {
-    ignorelist_set_invert(ignorelist, !IS_TRUE(value));
-  } else if (strcasecmp("NotifySensorAdd", key) == 0) {
-    c_ipmi_notify_add = IS_TRUE(value);
-  } else if (strcasecmp("NotifySensorRemove", key) == 0) {
-    c_ipmi_notify_remove = IS_TRUE(value);
-  } else if (strcasecmp("NotifySensorNotPresent", key) == 0) {
-    c_ipmi_notify_notpresent = IS_TRUE(value);
-  } else if (strcasecmp("SELEnabled", key) == 0) {
-    c_ipmi_sel_enabled = IS_TRUE(value);
-  } else if (strcasecmp("SELClearEvent", key) == 0) {
-    c_ipmi_sel_clear_event = IS_TRUE(value);
-  } else {
-    return -1;
+  st = calloc(1, sizeof(*st));
+  if (st == NULL) {
+    ERROR("ipmi plugin: calloc failed.");
+    return NULL;
   }
 
-  return 0;
-} /* int c_ipmi_config */
+  st->name = strdup("main");
+  if (st->name == NULL) {
+    sfree(st);
+    ERROR("ipmi plugin: strdup() failed.");
+    return NULL;
+  }
 
-static int c_ipmi_init(void) {
-  int status;
+  st->ignorelist = ignorelist_create(/* invert = */ 1);
+  if (st->ignorelist == NULL) {
+    sfree(st->name);
+    sfree(st);
+    ERROR("ipmi plugin: ignorelist_create() failed.");
+    return NULL;
+  }
 
-  /* Don't send `ADD' notifications during startup (~ 1 minute) */
-  time_t iv = CDTIME_T_TO_TIME_T(plugin_get_interval());
-  c_ipmi_init_in_progress = 1 + (60 / iv);
+  st->sensor_list = NULL;
+  pthread_mutex_init(&st->sensor_list_lock, /* attr = */ NULL);
+
+  st->host = NULL;
+  st->connaddr = NULL;
+  st->username = NULL;
+  st->password = NULL;
+  st->authtype = IPMI_AUTHTYPE_DEFAULT;
+
+  st->next = NULL;
+
+  return st;
+} /* c_ipmi_instance_t *c_ipmi_init_instance */
+
+static void c_ipmi_free_instance(c_ipmi_instance_t *st) {
+  if (st == NULL)
+    return;
+
+  assert(st->next == NULL);
+
+  sfree(st->name);
+  sfree(st->host);
+  sfree(st->connaddr);
+  sfree(st->username);
+  sfree(st->password);
+
+  ignorelist_free(st->ignorelist);
+  pthread_mutex_destroy(&st->sensor_list_lock);
+  sfree(st);
+} /* void c_ipmi_free_instance */
+
+void c_ipmi_add_instance(c_ipmi_instance_t *instance) {
```

Comment:
what about using a llist_t for instances?

Reply:
Is there is a strong opinion/rules about this?
Why would you prefer a list instead of chain?

Reply:
There's no clear-cut advise/guideline for or against the generic linked list. I'd say that if you need to look up by (string) key or you'd need a double-linked list, then the generic implementation can abstract away some complexity.

HTH, —octo

Reply:
>I'd say that if you need to look up by (string) key or you'd need a double-linked list
There is no such need here.